### DISCUSSION TITLE
Can AgentClassLoader be optimized?

### DISCUSSION THREAD

**Discussion by Oliverwqcwrw** [apache/skywalking#8942] [General]
In `skywalking-java`
`org.apache.skywalking.apm.agent.core.plugin.loader.AgentClassLoader#DEFAULT_LOADER`

I'm not sure I'm right in thinking that it would be better to make it volatile to prevent other threads from returning incomplete objects

**Comment by wu-sheng** [comment]
I am not sure what you mean. This class loader would be initialized once in the main thread, no further thread would be created before, so no CPU cache condition.

**Comment by Oliverwqcwrw** [comment]
I wonder if any other thread will call `org.apache.skywalking.apm.agent.core.plugin.loader.AgentClassLoader#getDefault` during `initDefaultLoader`

**Reply by wu-sheng** [reply]
AFAIK, no, it is impossible.
